log4j-api/src/main/java/org/apache/logging/log4j/ScopedContext.java

```diff
@@ -0,0 +1,224 @@
+     */
+    public static Map<String, Renderable> getContext() {
+        Deque<Map<String, Renderable>> stack = scopedContext.get();
+        if (stack != null && !stack.isEmpty()) {
+            return Collections.unmodifiableMap(stack.getFirst());
+        }
+        return Collections.emptyMap();
+    }
+
+    private static void addScopedContext(Map<String, Renderable> contextMap) {
+        Deque<Map<String, Renderable>> stack = scopedContext.get();
+        if (stack == null) {
+            stack = new ArrayDeque<>();
+            scopedContext.set(stack);
+        }
+        stack.addFirst(contextMap);
+    }
+
+    private static void removeScopedContext() {
+        Deque<Map<String, Renderable>> stack = scopedContext.get();
+        if (stack != null) {
+            if (!stack.isEmpty()) {
+                stack.removeFirst();
+            }
+            if (stack.isEmpty()) {
+                scopedContext.remove();
+            }
+        }
+    }
+
+    /**
+     * Return a new ScopedContext.
+     * @return the ScopedContext.
+     */
+    public static ScopedContext newInstance() {
+        return newInstance(false);
+    }
+
+    /**
+     * Return a new ScopedContext.
+     * @param inherit true if this context should inherit the values of its parent.
+     * @return the ScopedContext.
+     */
+    public static ScopedContext newInstance(boolean inherit) {
+        return new ScopedContext(inherit);
+    }
+
+    /**
+     * Return the key from the current ScopedContext, if there is one and the key exists.
+     * @param key The key.
+     * @return The value of the key in the current ScopedContext.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> T get(String key) {
+        Renderable renderable = getContext().get(key);
+        if (renderable != null) {
+            return (T) renderable.getObject();
+        } else {
+            return null;
+        }
+    }
+
+    private final Map<String, Renderable> contextMap = new HashMap<>();
+
+    private ScopedContext(boolean inherit) {
+        Map<String, Renderable> parent = ScopedContext.getContext();
+        if (inherit && !parent.isEmpty()) {
+            contextMap.putAll(parent);
+        }
+    }
+
+    /**
+     * Add all the values in the specified Map to the ScopedContext being constructed.
+     *
+     * @param map The Map to add to the ScopedContext being constructed.
+     * @return the ScopedContext being constructed.
+     */
+    public ScopedContext where(Map<String, Object> map) {
+        map.forEach(this::addObject);
+        return this;
+    }
+
+    /**
+     * Adds a key/value pair to the ScopedContext being constructed.
+     * @param key the key to add.
+     * @param value the value associated with the key.
+     * @return the ScopedContext being constructed.
+     */
+    public ScopedContext where(String key, Object value) {
+        addObject(key, value);
+        return this;
+    }
```

Comment:
This kind of methods should return a new copy of `ScopedContext` with a new copy of the map, in my opinion.

Code like:

```java
ScopedContext ctx = ScopedContext.newInstance().where("userId", "lala");
ctx.where("spanId", "abcdef").run(...);
ctx.run(...);
```

will fail otherwise.

Reply:
Actually, it won't fail. It just won't behave the same way you are believing it should. You are thinking that the run method on the second line should get a different instance then the run method on the third line. As written they will both get the same context. That doesn't make it wrong. Just different.  Again, this is simply a difference of opinion over how the API should behave. I am fine having a discussion about that.

However, I do see one possible concern. Someone could co
```
final ScopedContext ctx = ScopedContext.newInstance().where("userId", "lala");
ctx.where("spanId", "abcdef").run(() -> {
    ...
    ctx.where("spanId", "12345");
)};
```

Obviously calling where from inside the run or call methods should not be allowed, but that is easily preventable.  Admittedly though, returning an immutable object makes this impossible to cause any harm. This alone might convince me it should be immutable, as much as I really dislike that.

Reply:
It occurs to me that making the where method return an immutable object makes newInstance worthless. The boolean is no longer needed to clone the previous context since the where method would always do that.

Reply:
I have made ScopedContext immutable